src/app/subscriptions/data/index.ts

```diff
@@ -1,3 +1,38 @@
-export function getIsConnected(subscriptions: { connectTime?: string, disconnectTime?: string }[] | undefined = []): boolean {
-  return subscriptions.length > 0 && [subscriptions[subscriptions.length - 1]].every((item) => item.connectTime?.length && !item.disconnectTime)
+import type {
+  Version as PartialVersion,
+} from '@/types/index.d'
+
+export type Version = PartialVersion
+export type Subscription = {
+  disconnectTime?: string
+  status: {
+    lastUpdateTime: string
+  }
+  version?: Version
+}
+export type SubscriptionCollection<T extends { version?: any }> = {
+  subscriptions: T[]
+  connectedSubscription?: T
+  version?: T['version']
+}
+export const SubscriptionCollection = {
+  fromArray: <T extends Subscription>(items?: T[]): SubscriptionCollection<T> => {
+    const subscriptions = Array.isArray(items) ? items : []
+    // make a copy of the subscriptions and then sort with the latest one first
+    // so we don't change the order of the original array, plus any finding we
+    // do will always gives us the latest subscription if there are multiple
+    const subs = subscriptions.slice()
```

Comment:
**Comment**: While it’s true that this will leave the order of the initial array intact, it doesn’t enforce immutability as `Array.prototype.slice()` will only create a shallow copy meaning that `subs[0].name = 'test'` will still mutate `items`.

In short, we could skip this for now if we do this for the objective of enforcing immutability.

Reply:
Yeah its not deep, but if we then want to add properties to subscriptions that should probably be cloned elsewhere.

Just to note the deep freezing that we are doing in https://github.com/kumahq/kuma-gui/pull/1910 should always tell us if we try to mutate anything, even if there are no tests for it specifically. It caught some mutating in a previous iteration of this code and I think there is more of that elsewhere, those tests will tell us once we get those in.

Al in all, its fine to add when if/we need it as the test approach in https://github.com/kumahq/kuma-gui/pull/1910 should fail if we try to mutate anything original without cloning it first.